### ISSUE TITLE
figure out a whitelist

### ISSUE THREAD

**Issue by Monarchz3ro** [Monarchz3ro/Silver#14]
right now, the way things work, the commands loader will on purpose skip any module that has any import statement within itself
now. figure out a whitelist for modules that are "safe" for the player to load
stuff like `json` or `numpy`
1. **NO** modules that allow a player to cheese the game, for example,
`os.remove(path/to/victim/computer) -> disallow os` 
2. **NO** modules that want the player to connect to the internet
`response = requests.get("www.myasshurts.com") -> disallow requests`

that'll be that, i think. unless you can think of more criteria

**Comment by OcelotWalrus** [comment]
Maybe just, when running commands, check if they've the `x` permission (executable) in the `ecosystem.json` file permission and if yes, let the player run it.

Because making a whitelist would be a good idea, but the player could always change it anyway.

**Comment by OcelotWalrus** [comment]
By the way, there's a way to load external scripts way easier than the way you do it. I could change it in my PR if you'd like.

**Comment by Monarchz3ro** [comment]
> Because making a whitelist would be a good idea, but the player could always change it anyway.

The player can change every single aspect of the shell environment because it is open source. They can simply create a nuke script that targets the exact absolute path to the NPC computer and `shutil.rmtree()` it. 
This is not the problem.

There is an *intended* route for the player to take, i.e, never touching anything *outside* of the current root directory. Because if they do not have the permissions to x a script they can simply go to the `ecosystem.json` file and change permissions there. 

The player at every point has the freedom to mess with the game, and at that point, whatever they do is something I have chosen to completely ignore. All modification above the root folder is not intended and counts as cheating. 
Cheating has negative connotation, but I want to give that ability to players who think the way I handle things is too hard for them. There is an intended route, and a route where you simply look at the passwords in the json file.

Ah, by the way, if you are wondering how I intend to have the player computer (PathOS) interact with other computers, they will have their own filesystems that are essentially clones of the player's kernel folders. And there will be a networking module to ensure everything sticks together.
 
I haven't had much time to make a guide. I have art commissions whose deadlines are creeping up very soon and I must finish them, and my exams are inching closer. But I will make one.

**Comment by Monarchz3ro** [comment]
> By the way, there's a way to load external scripts way easier than the way you do it. I could change it in my PR if you'd like.

Sure! Show me how.

**Comment by OcelotWalrus** [comment]
Makes sense to me now that you've explained ;)

**Comment by OcelotWalrus** [comment]
Also, I don't have any other ideas of modules to whitelist, but I'll tell you if I have any in the future.

**Comment by OcelotWalrus** [comment]
> > By the way, there's a way to load external scripts way easier than the way you do it. I could change it in my PR if you'd like.
> 
> Sure! Show me how.

Actually, I'm just going to make an issue so you can copy the code easily and commit it yourself.

**Comment by OcelotWalrus** [comment]
> > By the way, there's a way to load external scripts way easier than the way you do it. I could change it in my PR if you'd like.
> 
> Sure! Show me how.

Actually, the way I'd code it isn't really better; after thinking about it.

**Comment by OcelotWalrus** [comment]
Maybe a 'no import' rule is what's needed?

**Comment by Monarchz3ro** [comment]
a whitelist is better than a blacklist, because there are a LOT more modules we'd rather disallow than allow. not sure though - maybe this'll change in the future

**Comment by OcelotWalrus** [comment]
Also maybe I had a thought: maybe `with open(` statements should be disallowed too, because they bypass the ecosystem permissions.

**Comment by OcelotWalrus** [comment]
I think `with open(` statements should be also disabled in what you call 'modules'. Because it bypass the ecosystem permission. But that'd be apart from import checks.

**Comment by Monarchz3ro** [comment]
I was thinking of it, aye. I'll look into it.

**Comment by OcelotWalrus** [comment]
Talking about that, certain kernel functions should be 'locked', like the `self.__user` and `self.root_dir` ones for example. Something looks like this:

```python

for i in locked_sys_var:
    if f"self.{i} =" in script:
      # skip script
```

**Comment by Monarchz3ro** [comment]
![image](https://github.com/Monarchz3ro/Silver/assets/151715484/0dcc4485-b6a9-4ade-b64a-28690fd08ecc)

**Comment by Monarchz3ro** [comment]
two underscores before a variable's name lock it to the home script, i.e you can't read/write self.__user() outside of master.py

**Comment by Monarchz3ro** [comment]
![image](https://github.com/Monarchz3ro/Silver/assets/151715484/37ef372d-1f75-4397-99d5-e93d773bfc34)
trying to print it directly throws a traceback

**Comment by OcelotWalrus** [comment]
Cool! Didn't know about that!

**Comment by Monarchz3ro** [comment]
yeah, i intend to lockdown every critical variable in master.py eventually
